avatars/avatars.js

```diff
@@ -388,6 +395,605 @@ const _makeDebugMesh = (avatar) => {
+
+    // const jointRight_ankleRight_toe = physicsManager.addJoint(flatMeshes.Right_ankle, flatMeshes.Right_toe, 
+    //   avatar.modelBoneOutputs.Right_ankle.modelBoneEnd.clone().multiplyScalar(0.5), 
+    //   avatar.modelBoneOutputs.Right_toe.modelBoneEnd.clone().multiplyScalar(0.5).negate(), 
+    //   new THREE.Quaternion().setFromEuler(localEuler.set(0, 0, 0)), new THREE.Quaternion(), false);
+    // physicsManager.setJointMotion(jointRight_ankleRight_toe, PxD6Axis.eTWIST, PxD6Motion.eLIMITED);
+    // physicsManager.setJointTwistLimit(jointRight_ankleRight_toe,       -10 * DEG2RAD,      90 * DEG2RAD);
+
+    // rootScene.children[2].visible = false; // test: hide E tag.  
+  };
+  object.setFromAvatar = avatar => {
+    // avatar.modelBoneOutputs.Root.updateMatrixWorld();
+    // todo: why one frame lag, even with avatar.modelBoneOutputs.Root.updateMatrixWorld().
+    for (const k in flatMeshes) {
+      const modelBone = avatar.modelBoneOutputs[k];
+      const meshBone = flatMeshes[k]; // ragdollMesh's
+
+      modelBone.matrixWorld.decompose(localVector, localQuaternion, localVector2);
+
+      localQuaternion2.copy(localQuaternion).multiply(
+        modelBone.forwardQuaternion
+      );
+      if (k === 'Hips') {
+        meshBone.matrixWorld.compose(localVector, localQuaternion, localVector2);
+        meshBone.matrixWorld.decompose(meshBone.position, meshBone.quaternion, meshBone.scale);
+      } else {
+        // put bone at center of neighbor joints
+        localVector.add(
+          localVector3.set(0, 0, -meshBone.boneLength * 0.5)
+            .applyQuaternion(localQuaternion2)
+        );
+        meshBone.matrixWorld.compose(localVector, localQuaternion, localVector2);
+        meshBone.matrixWorld.decompose(meshBone.position, meshBone.quaternion, meshBone.scale);
+      }
+      physicsManager.setTransform(meshBone);
+      physicsManager.setVelocity(meshBone, identityVector);
+      physicsManager.setAngularVelocity(meshBone, identityVector);
+    }
+    object.updateMatrixWorld();
+  };
+  object.toAvatar = avatar => {
+
+    {
+      avatar.modelBoneOutputs.Hips.position.set(0, 0, 0);
+      avatar.modelBoneOutputs.Hips.quaternion.identity();
+
+      flatMeshes.Hips.matrixWorld.decompose(avatar.modelBoneOutputs.Root.position, avatar.modelBoneOutputs.Root.quaternion, avatar.modelBoneOutputs.Root.scale)
+    }
+  
+    // global quaternion diff to locoal quaternion diff formula:
+    // https://forum.unity.com/threads/subtracting-quaternions.317649/
+    // A * B * iB = A
+    // A * B = C
+    // C * iB = A
+    // A = C * iB
+    // https://www.euclideanspace.com/maths/algebra/realNormedAlgebra/quaternions/functions/index.htm
+    // // qa' * qb' = (qb*qa)' 
+    // (A * B)' = B' * A'
+    // (A * B)' = C'
+    // B' * A' = C'
+    // B' * A' * A = B'
+    // C' * A = B'
+    // B' = C' * A
+    // B = (C' * A)'   !!! 
+    {
```

Comment:
This complicated repetitive math should be contained in helper functions.

Reply:
Added and moved `getDiffQuaternion()` into util.js. https://github.com/webaverse/app/pull/2731/commits/c9e3acb4f28c348cddba72a2ccad59b39d2f5fd8